text/0000-flexible-process-types.md

```diff
@@ -0,0 +1,217 @@
+# Meta
+[meta]: #meta
+- Name: Flexible Process Types
+- Start Date: 2025-01-07
+- Author(s): dmikusa
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Presently a Buildpacks can add process types, but there is no way to coordinate across Buildpacks about process types.
+
+For example, if I have a Time Buildpacks that wants to add the `time` command to the beginning of every process type's start command so that we can see `time` info for the commands run, that is not presently possible unless you modify the Buildpacks that generate the original process types to include this information.
+
+# Definitions
+[definitions]: #definitions
+
+**Process Type** - A process definition set by a buildpack in [`launch.toml`](https://github.com/buildpacks/spec/blob/main/buildpack.md#launchtoml-toml)
+
+# Motivation
+[motivation]: #motivation
+
+- Make process types more flexible. 
+- Support the use case where a buildpack wants to wrap another buildpack's process type's command
+- Support the use case where a buildpack wants to modify the arguments of another buildpack's process type
+
+# What it is
+[what-it-is]: #what-it-is
+
+Presently, a buildpack can define a process type in `launch.toml` like this:
+
+```
+[[processes]]
+type = "web"
+command = ["my-app"]
+args = ["arg1", "arg2"]
+default = false
+working-dir = "/workspace"
+```
+
+It is not presently possible for a different buildpack that runs subsequently to modify this process type definition.
+
+As described in the Motivation section, there are cases where this would be helpful and this RFC aims to make this both possible buildpack authors and safe for buildpack users.
+
+The proposal is that any subsequent buildpack can modify a previous buildpacks' process types by defining a transformation in `launch.toml` with the same type. The transformation defines the changes to be made. This will allow the subsequent buildpack to change the command, args, working directory, and indicate a reason for the change. The reason may be logged by tooling to help users understand why changes are being made.
+
+Because a subsequent buildpack does not know what a previous buildpack has defined for the command, args, and working directory the following place holders may be used to reference those values:
+
+- `$CMD` - the original command as a list, fits into a TOML list
```

Comment:
If a developer needs to add a literal `$CMD` they need a way to do so, especially since we are working with such a broad ecosystem. It also cannot involve escaping as any possible escape sequence could be a perfectly valid input as well.

I think the only option would be allowing developers to re-map these somehow like:

```
[[transforms.remap]]
CMD = "$ACTUAL_CMD"
ARGS = "$CUSTOM_ARGS"
```

That would allow them to use a $CMD literal if needed:

```
command = ["time", "$ACTUAL_CMD", "--format={$CMD%s}"]
```

Reply:
No immediate need, more enumerating edge cases. Aliasing doesn’t need to go into a v1 schema, but I’m making a note that if someone does have this problem, escaping isn’t the right fix.